delta/service/src/main/scala/ch/epfl/bluebrain/nexus/delta/service/plugin/PluginsLoader.scala

```diff
@@ -0,0 +1,92 @@
+package ch.epfl.bluebrain.nexus.delta.service.plugin
+
+import ch.epfl.bluebrain.nexus.delta.sdk.error.PluginError
+import ch.epfl.bluebrain.nexus.delta.sdk.error.PluginError.MultiplePluginDefClassesFound
+import ch.epfl.bluebrain.nexus.delta.sdk.plugin.PluginDef
+import ch.epfl.bluebrain.nexus.delta.service.plugin.PluginsLoader.PluginLoaderConfig
+import com.typesafe.scalalogging.Logger
+import io.github.classgraph.ClassGraph
+import monix.bio.IO
+
+import java.io.{File, FilenameFilter}
+import scala.jdk.CollectionConverters._
+
+/**
+  * Class responsible for loading [[PluginDef]]s.
+  *
+  * It looks for jar files in the passed directories in [[loaderConfig]] and tries to load a [[PluginDef]]
+  * from each jar file found.
+  *
+  * @param loaderConfig [[PluginsLoader]] configuration
+  */
+class PluginsLoader(loaderConfig: PluginLoaderConfig) {
+  private val logger: Logger = Logger[PluginsLoader]
+
+  private val parentClassLoader = this.getClass.getClassLoader
+
+  /**
+    * Loads all the available [[PluginDef]] from each of the discovered jar files
+    */
+  def load: IO[PluginError, (PluginsClassLoader, List[PluginDef])] =
+    for {
+      jarFiles                   <- IO.delay(loaderConfig.directories.flatMap(loadFiles)).hideErrors
+      (pluginsDef, classLoaders) <- IO.traverse(jarFiles)(loadPluginDef).map(_.flatten.sortBy(_._1).unzip)
+      accClassLoader              = new PluginsClassLoader(classLoaders, parentClassLoader)
+    } yield (accClassLoader, pluginsDef)
+
+  private def loadPluginDef(jar: File): IO[PluginError, Option[(PluginDef, PluginClassLoader)]] =
+    for {
+      pluginClassLoader <- IO.delay(new PluginClassLoader(jar.toURI.toURL, parentClassLoader)).hideErrors
+      pluginDefClasses  <- IO.delay(loadPluginDefClasses(pluginClassLoader)).hideErrors
+      pluginDef         <- pluginDefClasses match {
+                             case pluginDef :: Nil =>
+                               IO.pure(Some(pluginClassLoader.create[PluginDef](pluginDef, println)()))
+                             case Nil              =>
+                               logger.warn(s"Jar file '$jar' does not contain a 'PluginDef' implementation.")
+                               IO.pure(None)
+                             case multiple         =>
+                               IO.raiseError(MultiplePluginDefClassesFound(jar, multiple.toSet))
```

Comment:
If we log and ignore when there is no `PluginDef` found, we should do the same when there are multiple.

Reply:
I think it is different. There could be any random jar file in the directory that gets scanned which we don't care about. That's fine to issue a warning and continue.

But if there is a jar that contains an implementation of `PluginDef`it is meant to be used as a plugin. And if it contains multiple, it is an error that should not let the application start, but fail.